Article Refactor

trainerbill opened the pull request:
Superceeds https://github.com/meanjs/mean/pull/800

@ilanbiala @rhutchison

ilanbiala commented on modules/articles/client/config/articles.client.routes.js:
```diff
@@ -12,22 +12,51 @@ angular.module('articles').config(['$stateProvider',
       .state('articles.create', {
         url: '/create',
         templateUrl: 'modules/articles/client/views/create-article.client.view.html',
+        controller: 'ArticlesController',
```
Shouldn't this be 'ArticlesCreateController' or something similar since we are doing that above?

trainerbill replied:
Well previously I had them all separate controllers but that got shot down.  I can see the need for a create controller if we don't want to pass in a blank article to resolve.  Then the ArticlesController would handle edit/view since it needs an article to resolve?  @rhutchison

trainerbill commented:
@rhutchison Angular Stylized

I will fix the tests once we finalize the code.  Do the tests need to follow any sort of guidelines?

ilanbiala commented:
I'm not a fan of the IIFEs. Is anyone of the belief that what we have inside is actually being initialized as a global variable and should definitely not be there. It seems unnecessary to me.

trainerbill commented:
@ilanbiala I think the point is to have a style guide and stick to it everywhere, whether or not it is functionally necessary.

codydaig commented:
I would second the IIDEs. Other than that, I've been following th John Papa style guide in my most recent project and love it.

ilanbiala commented:
@codydaig what benefit does the IIFE offer? I don't see us creating many global variables.

trainerbill commented:
@rhutchison @ilanbiala @lirantal @codydaig Can I get a code review here?  I fixed the tests and everything is looking pretty good to me.  thx

As for the IIFEs, I think it is best practice and it is in line with our style guide, regardless if it is needed functionally.

simison commented:
> I don't see us creating many global variables.

Other projects might, as it's a boilerplate. It's a good practise to pass on for others, especially for less experienced developers.

ilanbiala commented:
@simison the other thing is that I'm pretty sure these IIFes are set up incorrectly. The whole point is that there are 0 global variables. In [articles.client.module.js](https://github.com/trainerbill/mean/blob/ArticleRefactor/modules/articles/client/articles.client.module.js), `ApplicationConfiguration` isn't passed in like it should be through the IIFE, but it is used as a global, so I'm not even sure why we are just adding IIFEs to add them. I'm still not in support of it, but if we are going to do it at all, then it should be implemented such that there are *no* global variables.

mleanos commented:
>Other projects might, as it's a boilerplate. It's a good practise to pass on for others, especially for less experienced developers.

@simison With the architecture of this boilerplate, I don't see the need for IIFE's. As long as users of this project keep their project's architecture inline with this implementation, then there shouldn't be a need for us to introduce this particular best practice. I would even say it would be confusing to less experienced developers; whereas how it's setup now, it's pretty easy to understand & read the code.

If someone deviates from the current implementation, and finds that they need to implement this best practice, then they could do so. But it would be a conscious effort on their part, and they'd have a specific reason for doing so.

simison commented:
@ilanbiala @mleanos All valid points and good food for thought!

trainerbill commented:
@ilanbiala You are correct on the articles.client.module.js and I will change that.  If you find any others let me know.

I don't think we should deviate from a style guide that we chose to use...  I think we need to stick to it otherwise how is a dev gonna know what to do and what not to do?

trainerbill commented:
Would we be ok with angular being global?  Or are we passing that into the IIFE as well?

codydaig commented:
@ilanbiala I meant I agree with you. I don't like IIFEs at all. I dislike how your code starts indented vefore you even write your first line. As long as you concat before you minify and uglify, it doesn't make a difference.

codydaig commented:
Sounds like its been decided to not use IIFEs. @trainerbill go ahead and remove them. I'll be able to do a more through review later day as well.

trainerbill commented:
@codydaig I don't know why we would decide that.  If we are going to have a style guide we should really stick to it.  Picking and choosing leaves gray areas.  Where was the discussion and decision made?  I know of 3 people that support them.

codydaig commented:
#636

ilanbiala commented:
@trainerbill we are picking a style guide and we are doing what makes sense for the framework. A style guide is exactly that, a guide for what you should do. It isn't a rulebook that says do everything exactly as I say. IIFEs aren't necessary because of the way our app is set up, therefore why add the complication and extra code that does nothing.

trainerbill commented:
@ilanbiala fair enough.  Someone should make a MEAN.js styleguide then.  It makes it easier as a contributor to know up front what I need to do, rather than go through 50 code reviews for style issues.

ilanbiala commented on modules/articles/client/controllers/articles.client.controller.js:
```diff
@@ -1,84 +1,54 @@
-      if (!isValid) {
-        $scope.$broadcast('show-errors-check-validity', 'articleForm');
+function ArticlesController($scope, $state, Authentication, articleResolve) {
+  var vm = this;
```
What is vm?

codydaig replied:
view model. It's supposed to replace $scope.

simison replied:
@ilanbiala https://github.com/johnpapa/angular-styleguide#controlleras-with-vm

ilanbiala replied:
The only issue is we can't just change it in one place. Maybe we do that as migration to 1.5?

ilanbiala commented on modules/articles/client/views/create-article.client.view.html:
```diff
@@ -1,20 +1,20 @@
-<section ng-controller="ArticlesController">
+<section >
```
space

ilanbiala commented on modules/articles/tests/client/articles.client.controller.tests.js:
```diff
@@ -1,14 +1,14 @@
-'use strict';
-
 (function () {
```
iife

trainerbill replied:
Its already this way in master.  You want me to remove the IIFEs from the tests as well?

https://github.com/meanjs/mean/blob/master/modules/articles/tests/client/articles.client.controller.tests.js

ilanbiala replied:
Yes, we should be consistent.

ilanbiala commented on modules/articles/tests/client/articles.client.controller.tests.js:
```diff
@@ -1,14 +1,14 @@
     // Initialize global variables
     var ArticlesController,
       scope,
+      rootScope,
```
change to have $

codydaig replied:
@trainerbill I was referring to these variable changes. They any variable that has a $ in the client side code also needs to have it in the tests for consistency.

trainerbill replied:
@codydaig I am honestly not the best with tests.  I didn't write them to begin with and was just making them "work" with the new style changes.  What would you say to merging this PR and fixing the styles on the tests at a later time?  or do you want to submit a PR to update the tests to this PR?  I just don't know enough about jasmine to do this solo

codydaig replied:
@trainerbill I'll submit a PR to this PR for you.

ilanbiala replied:
@codydaig did you submit a PR for this? @trainerbill this really isn't writing tests, it's just renaming the variables.

ilanbiala commented on modules/articles/tests/client/list.articles.client.controller.tests.js:
```diff
@@ -0,0 +1,94 @@
+(function () {
```
IIFE

ilanbiala commented on modules/articles/tests/client/list.articles.client.controller.tests.js:
```diff
@@ -0,0 +1,94 @@
+  describe('Articles List Controller Tests', function () {
+    // Initialize global variables
+    var ArticlesListController,
+      scope,
```
$

codydaig replied:
@trainerbill Here as well.

ilanbiala commented on modules/articles/tests/client/list.articles.client.controller.tests.js:
```diff
@@ -0,0 +1,94 @@
+    // Initialize global variables
+    var ArticlesListController,
+      scope,
+      rootScope,
```
$

codydaig replied:
@trainerbill Here as well.

ilanbiala commented on modules/articles/tests/client/routes.articles.client.tests.js:
```diff
@@ -0,0 +1,184 @@
+'use strict';
+
+(function () {
```
IIFE

ilanbiala commented on modules/articles/tests/client/routes.articles.client.tests.js:
```diff
@@ -0,0 +1,184 @@
+  // Articles Controller Spec
+  describe('Articles Route Tests', function () {
+    // Initialize global variables
+    var scope,
```
$

codydaig replied:
@trainerbill Here as well.

ilanbiala commented on modules/articles/tests/client/routes.articles.client.tests.js:
```diff
@@ -0,0 +1,184 @@
+  describe('Articles Route Tests', function () {
+    // Initialize global variables
+    var scope,
+      rootScope,
```
$

codydaig replied:
@trainerbill Here as well.

rhutchison commented:
@ilanbiala @simison @trainerbill @mleanos @codydaig I am personally in favor of IIFEs as it promotes best practices, consistency, and isolation from 3rd party modules.

In areas where we are defining `function routeConfig() { .. }` for each of our modules, and we are minifying our app, I would think without IIFEs that we will have issues (I haven't confirmed my theory).

I do not think a naming convention like articlesRouteConfig, usersRouteConfig, etc is a cleaner solution.  In-fact, when you consider other common declarations, this may end up being more keystrokes than having to create the IIFE in the first place.

simison commented:
+1 @rhutchison

trainerbill commented:
@rhutchison +1 obviously

ilanbiala commented:
@simison before it seems like you supported my and @mleanos's points, but now you want IIFEs? @rhutchison I'd still change the name to be clearer because they shouldn't all be the same like that anyway IMO.

trainerbill commented:
@ilanbiala I actually think that the naming should stay the same.  As long as we are in an IIFE it doesn't really matter, and it just makes more sense.  routeConfig over articleRouteConfig etc.  Will make the generator easier as well as it is less to code for in that regard.  Please let me know when a decision has been made.

mleanos commented:
@ilanbiala @trainerbill @rhutchison @codydaig @simison I think we can move forward with this and get it merged. If there are still some lingering concerns, we can address them apart from this PR. The remaining concerns don't seem to be blockers for this. The issue with IIFE's has been decided on, and we can re-address it later; if needed.

trainerbill commented:
@mleanos I don't think that we have addressed all concerns.  If IIFE's are removed I am guessing our  functions in things like routeConfig are going to overlap.  I am all for getting this merged ASAP, so someone needs to let me know what I need to do.

ilanbiala commented:
@trainerbill I think we should try without IIFEs for now until we run into any issues.

trainerbill commented:
@ilanbiala So if I remove the IIFE's from the test will this be good to merge?  I do not want to remove the vm from the controller/view.

ilanbiala commented:
@trainerbill the vm really doesn't fit into the current codebase. We have to stay consistent. When we update Angular, we will update what makes sense according to new conventions and better practices.

trainerbill commented:
@ilanbiala So you want me to introduce scope back into the controller and remove the controllerAs syntax?

ilanbiala commented:
@trainerbill for now, we need to in order to keep the codebase consistent. It doesn't make sense to do it piecemeal and before even updating Angular.

mleanos commented:
We obviously need to decide on what aspects of the style guide we should use. However, for the sake of getting what we're in agreement with merged in, from this PR, I say let's leave out anything those out. We can address the issues of the style guide, and this discussion shouldn't hold up this PR. There are other PR's/Issues that are waiting on this to get finalized.

lirantal commented:
I agree with @mleanos about not mixing the two topics until we have a clear definition for the style guide, which is very much related IMO to a PR that @rhutchison should submit about adding ESLint so we can enforce the rules we want as well.

trainerbill commented:
Someone please start a list of what it will take to get this merged and I will change it.  I am tired of this PR.

codydaig commented:
Let's remove the IIFEs. Address the other changes that @ilanbiala and I made in the line notes above to get this PR in. Then the style guide needs to be addressed before any other major PRs are submitted. 

@lirantal @ilanbiala @rhutchison @trainerbill @mleanos @simison Any objections?

codydaig commented:
I deleted my last comment as I realized what would be a slightly better plan of attack in my mind:

Let's follow the JohnPapa style guide, no diversions from it. Then after this PR get,s merged in, we can address any specific changes that we want to make. 

@ilanbiala @lirantal @mleanos @simison @rhutchison @trainerbill Any objections?

mleanos commented:
@codydaig I'm not a fan of that approach. If we introduce styles that haven't been decided on then as a boilerplate, dev users could end up adopting something we go away from. 

I think it would be better to remove the style guide implementation in this PR, with the exception of the routes config, and revert back to follow the rest of the project. Once this PR is merged & we pin point exactly what we want to use from John Papa's guide, every contributor will understand exactly what styles need to be followed. 

I'm very cautious about introducing code (styles) that we may revert later.

codydaig commented:
@mleanos your approach makes sense. I'm down for that Apple of attack.

trainerbill commented:
@mleanos Feel free to submit a PR that only changes the route config portions.  I am not going to revert this entire PR back to that state, as there was way too much work put into fixing the tests among other things.  I am going to put the IIFEs back into this PR and style it exactly like the style guide.  Hopefully in the future we can agree on using the style guide as a rulebook and rebase/merge.  @rhutchison

ilanbiala commented:
@mleanos so no IIFEs or anything in this PR?

mleanos commented:
@ilanbiala That's what I think is best. So we could regroup, and decide on the styles we want. My original thinking was that the Style Guide integration was slated for a later release. 

However, it appears we're at an impasse. Perhaps, we should just figure out the styles asap & move on from this issue.

Lastly, I'm not *completely* opposed to going with @codydaig suggestion of merging this as is; with the styles. I've expressed my concern over that option, but whatever everyone else wants to do, I'll go along with.

ilanbiala commented:
@mleanos @codydaig a big part is avoiding fragmentation and inconsistency. Any PR has to conform to the current style guide, even if there is no formal one.

lirantal commented:
I agree with @ilanbiala, let's not make a festival of coding styles and try to keep with what we have in the majority of the project, even though we might be heading a different way. Otherwise we'll just contribute to more spaghetti code

mleanos commented:
> a big part is avoiding fragmentation and inconsistency. Any PR has to conform to the current style guide, even if there is no formal one.

My sentiment exactly. 

@lirantal I'm in agreement

**Full disclosure**: I've recently lessened my stance on IIFE's, after a discussion with @rhutchison. *BUT* only in the context of the other aspects of John Papa's angular style guide being implemented. I'm not 100% convinced IIFE's will be necessary, but I won't be against them. What is most important to me right now is the consistency of the project, and these considerations not holding things up any longer. This PR shouldn't combine issues, by introducing styles that we haven't decided on.

simison commented:
@ilanbiala @rhutchison @mleanos 
> I've recently lessened my stance on IIFE's, after a discussion with @rhutchison. BUT only in the context of the other aspects of John Papa's angular style guide being implemented. 

This is exactly my stance; if several parts of the style guide are implemented, it becomes much more important to implement IIFE's. If not, with the current structure they're indeed not that important.

+1 for opening a new issue for discussion.

rhutchison commented:
imho, every PR for this release (0.4.2) *MUST* follow the angular style guide.  We will not release (0.4.2) until the project is consistent.  Once all of the features for this release have been merged, we will have a single PR to address anything that has not been addressed.

It is not fair to expect that 1 person take on the burden of re-writing everything in 1 PR to convert the entire app to conform to the style guide.  If the open PRs do not follow this rule, then everyone will before forced to follow the style-guide at some point - why not now?

ilanbiala commented:
@rhutchison people do use master code as well, so when they have inconsistent code it doesn't help them.

trainerbill commented:
@ilanbiala We discussed that when we decided to move to committing to the master branch.  It is considered a work in progress until a git tag specifies a release.  There is no way we are going to get one PR to reformat the whole project.

@rhutchison I definitely agree.  Once we get this first PR merged it is not going to take long at all to reformat the other modules.  We just need this as the baseline.

rhutchison commented:
If they're using master, they are using "latest".

rhutchison commented on modules/articles/client/config/articles.client.config.js:
```diff
@@ -23,5 +27,5 @@ angular.module('articles').run(['Menus',
     });
-  }
-]);
+  }  
```
remove extra spaces at the end of line

rhutchison commented on modules/articles/client/config/articles.client.routes.js:
```diff
@@ -1,36 +1,71 @@
-          roles: ['user', 'admin']
-        }
-      });
+    .state('articles', {
```
fix indentation

rhutchison commented on modules/articles/client/controllers/articles.client.controller.js:
```diff
@@ -1,58 +1,34 @@
+  function ArticlesController($scope, $state, Authentication, articleResolve) {
+    var vm = this;
+    vm.authentication = Authentication;
+    vm.article = articleResolve;
```
fix to be in alphabetical order

rhutchison commented on modules/articles/client/controllers/articles.client.controller.js:
```diff
@@ -1,58 +1,34 @@
-    // Update existing Article
-    $scope.update = function (isValid) {
-      $scope.error = null;
+    save.$inject = ['isValid'];
```
Is this necessary for tests? not sure why it's here.

rhutchison commented on modules/articles/client/services/articles.client.service.js:
```diff
@@ -11,4 +16,4 @@ angular.module('articles').factory('Articles', ['$resource',
       }
```
Is this extra object needed?  I think it can be removed.

trainerbill commented:
This PR is up to date with the style guide, checks passing, and ready to merge if we decide to go with it.

trainerbill commented:
any comments here?  Ready to merge?  LGTM's?

ilanbiala commented:
I'm not really a fan of these style guide conventions, so I'm going to defer until others chip in.

rhutchison commented:
I know it was a painful commit/discussion - Thank you for working through it.

LGTM

codydaig commented:
@trainerbill There are a couple variable declarations that @ilanbiala pointed out in line comments that need to get fixed. Also, can you mark the tests as pending instead of commenting them out so they show up as needing attention? `xit('', function () {})` instead of `it('', function () {})`

trainerbill commented:
@codydaig Done. Rebase.  Traveling the next 2 weeks so my responses will be delayed if anything else needs to change.

codydaig commented on modules/articles/tests/client/articles.client.controller.tests.js:
```diff
@@ -1,14 +1,14 @@
   describe('Articles Controller Tests', function () {
     // Initialize global variables
     var ArticlesController,
       scope,
```
`$scope`

codydaig commented on modules/articles/client/services/articles.client.service.js:
```diff
@@ -1,8 +1,13 @@
+
+  Articles.$inject = ['$resource'];
+
+  function Articles($resource) {
```
@trainerbill To follow John Papa to the T, this should be ArticlesService

https://github.com/johnpapa/angular-styleguide#naming

trainerbill replied:
You are correct sir.  I will update

ilanbiala replied:
Need to update this

mleanos replied:
This still hasn't been updated.

ilanbiala replied:
update

lirantal commented:
@trainerbill this is for v0.4.2, can you please update and get this ready?
@ilanbiala any other changes required here for @trainerbill or can we move on with this for an v0.4.2 release?

ilanbiala commented:
@lirantal since this has some style guide stuff in this, we should probably finalize that first before merging this in to avoid refactoring it later.

lirantal commented:
Ok so maybe set milestone to 0.5?

trainerbill commented:
@lirantal This PR is ready.

lirantal commented:
Cool. @ilanbiala's got it.

codydaig commented:
@trainerbill Can you update the last of the changes. Then I think this is ready.

trainerbill commented:
@codydaig Haven't looked at this in a long time.  I think it is ready.  What do you think needs to change?

ilanbiala commented:
Commit message needs to be fixed at least.

codydaig commented:
There's issues in line comments specifying certain variable names and naming of services.

mleanos commented:
@trainerbill I just went through the diff notes, and there are some unaddressed issues... 

For instance, https://github.com/meanjs/mean/pull/874/files#r40753755
https://github.com/meanjs/mean/pull/874/files#r38559922 & a few others.

Can you through the comments and double check? Also, it might be a good idea to do a rebase, when you update the commit message, and make sure everything is working still. Just a quick smoke check would be fine.

ilanbiala commented on modules/articles/client/controllers/articles.client.controller.js:
```diff
@@ -1,84 +1,51 @@
-        return false;
-      }
+  function ArticlesController($scope, $state, Authentication, articleResolve) {
+    var vm = this;
```
why is this necessary?

ilanbiala commented on modules/articles/client/controllers/list.articles.client.controller.js:
```diff
@@ -0,0 +1,15 @@
+  ArticlesListController.$inject = ['Articles'];
+
+  function ArticlesListController(Articles) {
+    var vm = this;
```
?

ilanbiala commented:
@trainerbill the vm concept was supposed to be removed because we haven't updated Angular yet, which should happen after this PR is merged, and will be the last thing we do with no new code being added so we can see exactly what happens when we update the code. @lirantal @codydaig do you want  to update to Angular 1.5 for 0.5.0 and release fewer features than we normally would so there is no more delay?

mleanos commented:
@trainerbill What was decided on how to handle a 400 (Bad Request) with this new UI route implementation?

For instance, when an invalid `:articleId` is provided, this will receive an error from the api call & not get handled on the client. Same for when the article isn't found.

mleanos commented:
@ilanbiala Why is updating Angular a factor in using the `var vm = this;` concept?

I've been using it in a MEANJS project for quite some time, and I don't see the relevance to what version of Angular I'm using.

ilanbiala commented:
@mleanos if I'm not mistaken, that line is for using controller as syntax, which we won't be doing piecemeal. We'll change whatever makes sense in terms of Angular implementation when we update Angular itself.

codydaig commented:
@ilanbiala I don't want to upgrade to 1.5 yet as it's still in beta. I'm not opposed to bumping to 1.4 which is only one minor version bump from where we are currently (1.3) and i don't think using the controller syntax needs to wait until the version bump.

ilanbiala commented:
@codydaig I think 1.5.0 will be released in the next month or so. They have already released rc.0, and they took just over a month to release 1.4.0 from that point. I'm fine with using controller as, but I'd prefer we don't just do the article controllers and then do the rest later piecemeal, because that usually doesn't work well.

lirantal commented:
how about we wait, and then upgrade to angular 2?

mleanos commented:
I don't see problem having the *controllerAs* syntax in just the Articles module. I thought our plan was to roll out the Angular Style-guide one module at a time; as to not complicate PR's. For instance, in my #807 , I was planning on implementing the Angular Style-guilde; including the `controllerAs` on the states. 

Using this syntax in one module, has no affect on any other module.

mleanos commented:
@lirantal What are you suggesting we wait on?

Using the `controllerAs` syntax? Or just upgrading Angular?

My vote would be to just wait on upgrading Angular, unless we see a benefit of doing so earlier. Then once we're ready, we upgrade to Angular 2.

lirantal commented:
it was part joke, part serious.
is the effort in upgrading to newer versions of Angular (which are not the angular2 version) worth it?

ilanbiala commented:
Honestly, I don't see Angular 2 as very well accepted right now considering all the fundamental changes, and Angular 1.x still has a lot of improvements over 1.3, so I see a point in upgrading. @mleanos I'm for the syntax, but not piecemeal, because then we release that because we forget and then users have fragmented code.

mleanos commented:
@ilanbiala I think we can get each module updated to use the new Angular Style-guide, as part of the 0.5.0 release; as was the plan anyway. Having "fragmented code" doesn't pose any issues for our users, if they get this implementation in piecemeal anytime between now & when we release 0.5.0 

I don't think going at every module for the Angular Style-guide implementation is wise. That PR would be too big, and the contributor may face problems that are unique to one module; thus, blocking the changes for the other modules that shouldn't have anything to do with the module posing the issue.

Let's continue with the plan to introduce John Papa's Angular Style-guide as part of 0.5.0. There's a lot of benefit in the style-guide, and it will help keep the framework clean (and DRY) as we implement more architectural changes; for instance, Token-based Auth. Not to mention the guide advocates for best practices.

mleanos commented:
As for whether or not to upgrade Angular.. I think upgrading to Angular 2 won't happen anytime soon for this framework; we should wait until it's more mature & we have the bandwidth to handle that upgrade.

I don't really have much of an opinion about upgrading to Angular 1.5. I don't see it introducing any drastic changes that would affect what we're doing on the front-end. And still, I don't see the relevance of what version of Angular we need to use `controllerAs` in our states.

mg1075 commented:
+1 for `controllerAs` syntax in all modules, and not piecemeal. 
That seems like too much overhead if some modules implement it, while others do not. It adds to the concept count instead of reducing it. The truly "core" modules of the boilerplate are `core` and `users`, and those should have controllerAs implemented before anything else. But once those are done, switching Articles to controllerAs would, I think, go more smoothly. At least you would know all modules are trying to adhere to John Papa's guidelines, instead of only some.

mleanos commented:
@mg1075 How does having `controllerAs` create overhead for other modules?

I think introducing John Papa's Angular style-guide in Articles first, is the best course of action. It's our basic CRUD module, and is provided as an example to go off of. It's more important for us to have that precedent over the other modules.

ilanbiala commented:
@mleanos Angular 1.5 has much better performance, introduces some breaking changes (both good and bad), and some new features that users will definitely benefit from. I think upgrading it will help. I'm all for the style guide, but I'd rather do every feature all at once rather than each feature piecemeal, because inevitably we're gonna miss a little bit of legacy code.

mg1075 commented:
@mleanos - I think there would be overhead if `controllerAs` is not used in **all** modules. 

If `controllerAs` is only used in one or two modules to start off, now a person has to start switching between patterns: this view and controller use `controllerAs`, this view and controller don't; this view and controller follow the angular style guidelines, this other view and controller don't.

One advantage of using `controllerAs` noted in the guidelines is: 
> Why?: Helps avoid using $parent calls in Views with nested controllers.

Imagine I have an app where there is an angular Controller that is a parent controller, sort of like a shell controller for an app. And then, of course, I have a controller nested inside. Using `controllerAs` means I won't have to use `$parent`, and in a framework like MEAN.JS where - in my experience - you have to jump around from place to find things, avoiding the use of `$parent` will make the code easier to read. But this is unlikely to happen unless `core` and `users` use `controllerAs`.

mleanos commented:
@mg1075 I'm advocating to use `controllerAs`. But I would rather avoid tasking one contributor with implementing this style into every module in one PR. I would rather see this done by multiple contributors, in a shared branch; which is what I view *master* at this point, since our next release is a major bump to 0.5.0.

IMO, most users of this framework don't modify the *core*, or *users* module. They take our example CRUD (Articles) and use it as a guide for developing their own custom modules. This is another reason why it makes more sense to implement the style-guide into the Article module to start.

mleanos commented:
@mg1075 I'm on Gitter right now, if you'd like to discuss this there. To avoid cluttering this issue.

codydaig commented:
@mleanos @mg1075 @lirantal @ilanbiala I'd much rather see it broken up by module rather than one large PR. Because that PR is going to be a nightmare for anyone doing it if any changes are made along the way. All modules need to be switched over before 0.5 is released though to avoid overhead. The master branch is considered development, so anyone using it in production runs the risk of some type of overhead as not all proposed changes will have been made yet.

ilanbiala commented:
@codydaig do you want to oversee that then? It's fine if people submit PRs too, but if not then would you be able to refactor that code? I'll gladly review it as well to help out.

Also, one thing that I'd want to see if we're doing it now is avoiding $scope as much as possible. https://toddmotto.com/no-scope-soup-bind-to-controller-angularjs/ http://blog.ninja-squad.com/2015/07/21/what-is-new-angularjs-1.4/ have suggestions on how to do that, but they require Angular 1.4.

codydaig commented:
@ilanbiala I can oversee that. I've worked with the John Papa style guide. (Actually more familiar with that then using $scope.) I agree, $scope shouldn't have to be used anywhere with the ControllerAs syntax. I will try upgrading Angular and see if it breaks anything. These next 2 weeks are crazy for me, but come January I should be able to crank these out.

mg1075 commented:
@codydaig - recall @rhutchison comments in gitter about using controllerAs:

> it seems to be working well, I am still adopting it in another project
> we change to best-practice as we touch files
> only thing that sucks is though **may not be able to remove $scope from DI
> for things like $scope.$apply(), $scope.$emit(), $scope.$watch, etc**.
> but controllerAs is working well

...and...

> one thing I came across today, is in instances like a menu or tab system, where I use 
> templateProvider/controllerProvider on my routes to dynamically set my controller/templates, there 
> is no controllerAsProvider, so I need to define my controller like so `HomeController as vm`
> example
```
function controllerProvider($stateParams, companyTabs) {
                    if (companyTabs[$stateParams.tab].controller) {
                        var controller = companyTabs[$stateParams.tab].controller.split(' as ');

                        if (companyTabs[$stateParams.tab].controllerAs) {
                            controller[1] = companyTabs[$stateParams.tab].controllerAs;
                        }

                        return controller.join(' as ');
                    }
                }],
```

mleanos commented:
I'd like to undertake the refactoring of the Users Admin feature. I have a good amount of experience with John Papa's styleguide as well. 

As for the use of `$scope`, we don't need it in any controllers. For anything that needs to use it like `$broadcast`, `$emit`, etc., then that functionaly should be abstracted out into services, and/or directives.

mleanos commented:
@trainerbill @rhutchison what's the plan here?

rhutchison commented:
@trainerbill clean up the commit message and squash the commits - LGTM

mleanos commented on modules/core/client/app/init.js:
```diff
@@ -19,7 +19,7 @@ angular.module(ApplicationConfiguration.applicationModuleName).run(function ($ro
       var allowed = false;
       toState.data.roles.forEach(function (role) {
-        if (Authentication.user.roles !== undefined && Authentication.user.roles.indexOf(role) !== -1) {
+        if ((role === 'guest') || (Authentication.user && Authentication.user.roles !== undefined && Authentication.user.roles.indexOf(role) !== -1)) {
```
@rhutchison Why is this included? It's already in master.

We should be using rebase, not merges right?

ilanbiala commented on modules/articles/client/config/articles.client.routes.js:
```diff
@@ -12,25 +16,59 @@ angular.module('articles').config(['$stateProvider',
+
+  newArticle.$inject = ['Article'];
+
+  function newArticle(Article) {
```
Why do this? Why not just do it in the controller

trainerbill replied:
@ilanbiala It makes the controller smaller and allows the use of UI-Router nested views to share the resolve between the views and other controllers.

mleanos commented on modules/articles/client/services/models/article.client.model.js:
```diff
@@ -0,0 +1,19 @@
+(function () {
```
@rhutchison What's the purpose of this file? 

We haven't decided on how we're going to move to a client-side model implementation. This isn't needed, since it's just a duplication of *article.client.service*

mleanos commented:
@trainerbill There's a reference to a merge conflict still in the code.. I commented on the line that has `<<<<<<<< HEAD ...`

Also, there are some older line comments that haven't been addressed.
